src/GafferImage/Display.cpp

```diff
@@ -228,22 +221,60 @@ class GafferDisplayDriver : public IECoreImage::DisplayDriver
 			return true;
 		}
 
-		ConstFloatVectorDataPtr channelData( const Imath::V2i &tileOrigin, const std::string &channelName )
+		ConstFloatVectorDataPtr channelData( const Imath::V2i &tileOrigin, const std::string &channelName, int dataCount )
 		{
 			vector<string>::const_iterator cIt = find( channelNames().begin(), channelNames().end(), channelName );
 			if( cIt == channelNames().end() )
 			{
 				return ImagePlug::blackTile();
 			}
 
-			ConstFloatVectorDataPtr tile = getTile( tileOrigin, cIt - channelNames().begin() );
-			if( tile )
+			int tileIndex = getTileIndex( tileOrigin, cIt - channelNames().begin() );
+			if( tileIndex == -1 )
 			{
-				return tile;
+				return ImagePlug::blackTile();
 			}
-			else
+
+			TileCacheElement &tileData = m_tiles[ tileIndex ];
+
+			int previousDataCount;
+
 			{
-				return ImagePlug::blackTile();
+				tbb::spin_rw_mutex::scoped_lock tileLock( m_tileMutex, false /* read */ );
+
+				previousDataCount = tileData.cachedForDataCount;
+
+				if( tileData.cachedForDataCount == dataCount || !tileData.dirty)
+				{
+					// In order to ensure hashes and computes are consistent, once we have
+					// bound a version of the tile for this dataCount, we don't change it,
+					// even if it has been dirtied in the meantime.
+					return tileData.cachedTile;
+				}
+			}
+
+			FloatVectorDataPtr newCache = new FloatVectorData();
+			newCache->writable() = tileData.backBuffer;
+
+			{
+				tbb::spin_rw_mutex::scoped_lock tileLock( m_tileMutex, true /* write */ );
+
+				if( tileData.cachedForDataCount != previousDataCount && tileData.cachedForDataCount >= dataCount )
+				{
+					// Another process has beaten us to write the cache, and their data count was >= ours,
```

Comment:
How common is this? It seems fairly plausible that we could hit it pretty hard when hashing a downstream Resize or Blur, where many downstream tiles will rely on the same upstream tiles. In the worst case, could every thread end up doing a copy and then all but one doing a discard? This seems like maybe it has the potential to be worse than the copy-per-renderer-bucket that we were doing in the old code, in certain situations?